### PR TITLE
Add tagging system to tickets, and auto reply to emails functionality 

### PR DESCRIPTION
**Pull request by jahnaviraj**

### CODE REVIEW

File: frappedesk/api/ticket.py [hunk 1]
```diff
@@ -139,6 +139,40 @@ def assign_ticket_type(ticket_id, type):
...
+		ticket_doc = frappe.get_doc("Ticket", ticket_id)
+
+		tag_exists = False
+		for ticket_tag in ticket_doc.ticket_tag:
+			if ticket_tag.tag == tag:
+				tag_exists = True
+
+		if not tag_exists:
+			check_and_create_ticket_tag(tag)
+			ticket_doc.append("ticket_tag", {
```
Comment (ashishnpbs) [inline frappedesk/api/ticket.py:154]: only if create is success

File: frappedesk/api/ticket.py [hunk 2]
```diff
@@ -139,6 +139,40 @@ def assign_ticket_type(ticket_id, type):
...
+
+		return ticket_doc
+
+@frappe.whitelist(allow_guest=True)
+def delete_ticket_tag(ticket_id, tag):
+	ticket_doc = frappe.get_doc("Ticket", ticket_id)
+
+	for row in ticket_doc.ticket_tag:
+		if row.name == tag:
+			ticket_doc.remove(row)
```
Comment (ashishnpbs) [inline frappedesk/api/ticket.py:168]: remove from ticket_doc or ticket_doc.ticket_tag

ideally remove should not be part of for loop -- you can check in the loop and if exists remove

File: frappedesk/api/ticket.py [hunk 3]
```diff
@@ -139,6 +139,40 @@ def assign_ticket_type(ticket_id, type):
...
+			ticket_doc.save()
+			break
+
+	log_ticket_activity(ticket_id, f"tag: {tag} removed.")
```
Comment (ashishnpbs) [inline frappedesk/api/ticket.py:172]: only if tag is removed

File: frappedesk/api/ticket.py [hunk 4]
```diff
@@ -260,6 +294,17 @@ def check_and_create_ticket_type(type):
 
 	return ticket_type_doc
 
+@frappe.whitelist(allow_guest=True)
+def check_and_create_ticket_tag(tag):
+	if not frappe.db.exists("Helpdesk Tag", tag):
```
Comment (ashishnpbs) [inline frappedesk/api/ticket.py:299]: don't have a function call inside if condition - call the function first and put the condition on return value

File: frappedesk/frappedesk/doctype/helpdesk_tag/helpdesk_tag.json [hunk 5]
```diff
@@ -0,0 +1,46 @@
+{
```
Comment (ashishnpbs) [inline frappedesk/frappedesk/doctype/helpdesk_tag/helpdesk_tag.json:1]: do we need to give permission to HelpDesk Agent and Manager to create tags

File: desk/src/components/desk/ticket/ActionPanel.vue [hunk 6]
```diff
@@ -171,6 +171,69 @@
...
+							</div>
+							</div>
+					</div>
+					<Autocomplete 
+						v-if="ticketTags"
+						:options="ticketTags.map(x => {
+							return {label: x.name , value: x.name}
+						})"
+						placeholder="Set tags"
+						:value="ticket.ticket_tag.length > 0  && !updatingTicketType ? ticket.ticket_tag[0].name : ''"
```
[the remark this answers is not in the document]
Comment (jahnaviraj) [reply desk/src/components/desk/ticket/ActionPanel.vue:199]: Updated

File: desk/src/components/desk/ticket/ActionPanel.vue [hunk 7]
```diff
@@ -171,6 +171,69 @@
 						</template>
 					</Autocomplete>
 				</div>
+				<div class="flex flex-col space-y-[8px]" :class="mandatoryFieldsNotSet && !ticket.ticket_tag ? 'error-animation' : ''">
```
[the remark this answers is not in the document]
Comment (jahnaviraj) [reply desk/src/components/desk/ticket/ActionPanel.vue:174]: removed

File: desk/src/pages/desk/Desk.vue [hunk 8]
```diff
@@ -401,6 +423,22 @@ export default {
...
+				params: {
+					doctype: 'Helpdesk Tag',
+					pluck: 'name',
+				},
+				auto: this.user.has_desk_access,
+				onSuccess: (data) => {
+					this.ticketTags = data
+				},
+				onError: () => {
+					// TODO:
```
[the remark this answers is not in the document]
Comment (jahnaviraj) [reply desk/src/pages/desk/Desk.vue:438]: added everywhere where ever missing

File: frappedesk/api/ticket.py [hunk 9]
```diff
@@ -139,6 +139,40 @@ def assign_ticket_type(ticket_id, type):
 
 		return ticket_doc
 
+@frappe.whitelist(allow_guest=True)
+def assign_ticket_tag(ticket_id, tag):
+	if ticket_id:
+		ticket_doc = frappe.get_doc("Ticket", ticket_id)
+
```
[the remark this answers is not in the document]
Comment (jahnaviraj) [reply frappedesk/api/ticket.py:146]: updated

File: frappedesk/api/ticket.py [hunk 10]
```diff
@@ -139,6 +139,40 @@ def assign_ticket_type(ticket_id, type):
...
+		tag_exists = False
+		for ticket_tag in ticket_doc.ticket_tag:
+			if ticket_tag.tag == tag:
+				tag_exists = True
```
[the remark this answers is not in the document]
Comment (jahnaviraj) [reply frappedesk/api/ticket.py:150]: updated

File: frappedesk/api/ticket.py (hunk 1 above)
Comment (jahnaviraj) [reply frappedesk/api/ticket.py:154]: Updated

File: frappedesk/api/ticket.py (hunk 2 above)
Comment (jahnaviraj) [reply frappedesk/api/ticket.py:168]: Updated

File: frappedesk/api/ticket.py (hunk 3 above)
Comment (jahnaviraj) [reply frappedesk/api/ticket.py:172]: Done

File: frappedesk/api/ticket.py (hunk 4 above)
Comment (jahnaviraj) [reply frappedesk/api/ticket.py:299]: Done

File: frappedesk/frappedesk/doctype/helpdesk_tag/helpdesk_tag.json (hunk 5 above)
Comment (jahnaviraj) [reply frappedesk/frappedesk/doctype/helpdesk_tag/helpdesk_tag.json:1]: Added

**Pull request merged**